### PR TITLE
Enable distributed training on SageMaker for custom training job task 

### PR DESCRIPTION
**Pull request by bnsblue**
# TL;DR
This PR adds a field in TrainingJobResourceConfig in Flytekit to allow users to specify the protocol (e.g., MPI) they want to use to train their job in a distributed manner.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This PR adds a field in TrainingJobResourceConfig in Flytekit to allow users to specify the protocol (e.g., MPI) they want to use to train their job in a distributed manner.

## Tracking Issue
https://github.com/lyft/flyte/issues/452

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -23,7 +23,8 @@
         ]
     },
     install_requires=[
-        "flyteidl>=0.18.8,<1.0.0",
+        #"flyteidl>=0.18.8,<1.0.0",
+        "flyteidl @ git+ssh://git@github.com/lyft/flyteidl@2efbce3a933f6c00a3c7933a07a031a091e861d7#egg=flyteidl"
```
Comment (wild-endeavor) [inline setup.py:27]: merge idl first?

Comment (kumare3) [reply setup.py]: yup

**Comment by kumare3** [comment]
LGTM

File: setup.py (hunk 1 above)
Comment (bnsblue) [reply setup.py]: fixed

File: Makefile [hunk 2]
```diff
@@ -36,6 +36,10 @@ test: lint ## Run tests
 	pytest tests/scripts
 	shellcheck **/*.sh
 
+requirements-spark3.txt: export CUSTOM_COMPILE_COMMAND := make requirements-spark3.txt
```
Comment (wild-endeavor) [inline Makefile:39]: why is this showing up?

File: flytekit/models/sagemaker/training_job.py [hunk 3]
```diff
@@ -43,6 +58,15 @@ def volume_size_in_gb(self) -> int:
         """
         return self._volume_size_in_gb
 
+    @property
+    def distributed_protocol(self) -> int:
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/training_job.py:62]: Can you add this property to the unit test for the model as well please?

File: Makefile (hunk 2 above)
Comment (bnsblue) [reply Makefile:39]: I made a change to setup.py, so needed to update requirements-spark3.txt otherwise the python3.8 check wouldn't pass

File: flytekit/models/sagemaker/training_job.py (hunk 3 above)
Comment (bnsblue) [reply flytekit/models/sagemaker/training_job.py:62]: Added.

**Comment by bnsblue** [comment]
Just noticed that I didn't add unit tests for SageMaker models so I added a set of them. When writing the unit tests I did find a bug in `ParameterRanges.from_flyte_idl()`.

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 4]
```diff
@@ -177,11 +178,17 @@ def to_flyte_idl(self) -> _idl_parameter_ranges.ParameterRanges:
...
         for k, v in pb2_object.parameter_range_map.items():
-            if isinstance(v, _idl_parameter_ranges.ContinuousParameterRange):
-                converted[k] = ContinuousParameterRange.from_flyte_idl(v)
-            elif isinstance(v, _idl_parameter_ranges.IntegerParameterRange):
-                converted[k] = IntegerParameterRange.from_flyte_idl(v)
+            # Use WhichOneof to find out which field in the oneof has been set
+            # https://developers.google.com/protocol-buffers/docs/reference/python-generated#oneof
+            # and then use getattr to get the field, and use isinstance to check if that field is indeed that type
+            # https://stackoverflow.com/questions/51421392/dynamically-access-oneof-value-from-a-protobuf
+            if isinstance(getattr(v, v.WhichOneof('parameter_range_type')),
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py]: I thought we usually used HasField for this stuff.

Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py]: fixed :)

**Review by wild-endeavor — COMMENTED** [review]
last comment then good to go

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 5]
```diff
@@ -152,7 +152,8 @@ def to_flyte_idl(self) -> _idl_parameter_ranges.CategoricalParameterRange:
 
     @classmethod
     def from_flyte_idl(cls, pb2_object: _idl_parameter_ranges.CategoricalParameterRange):
-        return cls(values=pb2_object.values)
+        #    return cls(values=pb2_object.values)
```
[the remark this answers is not in the document]
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py]: Without doing that, this following test would fail
```
def test_categorical_parameter_range():
    case = unittest.TestCase()
    pr = parameter_ranges.CategoricalParameterRange(values=["abc", "cat"])
    pr2 = parameter_ranges.CategoricalParameterRange.from_flyte_idl(pr.to_flyte_idl())
    case.assertEqual(type(pr2.values), list)
```
with error message
```
  File "/Users/changhonghsu/src/flytekit/tests/flytekit/unit/models/sagemaker/test_parameter_ranges.py", line 40, in test_categorical_parameter_range
    case.assertEqual(type(pr2.values), list)
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pycharm/teamcity/diff_tools.py", line 38, in _patched_equals
    raise error
teamcity.diff_tools.EqualsAssertionError:  :: <class 'google.protobuf.pyext._message.RepeatedScalarContainer'> != <class 'list'>
----------------------------------------------------------------------
Ran 1 test in 0.034s
FAILED (failures=1)
```

**Review by wild-endeavor — APPROVED** [review]
